hmc-cradlepoint/cradlepoint
# Fix BOM

## Things to test for:
- Adding an item to a BOM in testCase level
  - make sure you can enter quantity (need to double-click on the cell to edit. after you are editing a specific cell, need to click somewhere else so that the change is saved) and whether it isOptional is the modal
  - make sure summaryBOM updates properly (only updates if quantity>before)
  - make sure navigation between select device and select quantity modals is working
- Editing an item of a BOM in testCase level by changing the quantity
  - make sure summaryBOM updates properly 
- Cloning a test case and test plan
  - make sure the BOM is copied over 
  - make sure summaryBOM updates properly 

## Major files to look at:
(PS: I did some minor styling changes to many other files. Thus, I will list the files that actually had major logic changes that reviewer should focus on)
### Data schema
- `bomDeviceSchema`: this defines what adding a single item to the BOM or summaryBOM should look like

### Front-end
- `5TestCaseDetails`
     - mostly changed what props to pass in for the selectDevice and selectQuantity Modal
     - notes about the variable`selectedIDs` 
          - when we are adding a new device, this variable stores the ids of the devices (deviceId) that are selected from the `selectDevice` modal, and are then passed into the `selectQuantity` modal to populate the table
          - when we are editing a device from the current BOM, this variable stores the id of the item (_id) that we selected to edit, and pass **directly** to the `selectQuantity` modal (does not go through `selectDevice` modal)
- modals
    - `selectDevice.jsx`
         - allow users to select what devices they want to add by updating `selectedIDs`
    - `selectQuantity.jsx`
         - once it gets the `selectedIDs` variable, it will use it to generate `data` via the function `formatData(id) which is then used to populate the table
         - `formatData`: note behavior of edit and add are different
         - ToastContainer is just a pop-up to warn users that quantity they enter should be greater than 0

### Back-end
-  `addDeviceToBOM.js` (most of the logic done is actually in addEntry)
- `addEntry.js`:
    - `addDeviceToBOM(data)`: adds each item to the BOM and also updates the summaryBOM 
              - note that update the summaryBOM means that it will check every testCase in the testPlan and see if they have the same device AND isOptional setting. If a testCase contains such item, we need to compare the quantity and get the max. This way, summaryBOM will ALWAYS be displaying the maximum quantity of a specific device and setting among the test cases. 
- `editDeviceInBOM.js`
    - updates the BOM and summaryBOM 
- `cloneEntry.js`: 
    - `cloneTest(testCaseId, oldTests)` (helper function)
        - clones the test, given the original tests and the new testCase's id
    - `cloneBOM(testPlanId, oldBOM, client)` (helper function!)
        - returns a new BOM, copied over from the original BOM but each item has a new id
        - also updates summaryBOM of the test plan
    - `cloneTestCase(data)`
        - calls cloneBOM to handle cloning BOM
    - `cloneTestPlan(data)`
         -  calls cloneTestCase
